tf2onnx/handler.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
+"""Opset registry."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import collections
+import inspect
+
+from tf2onnx import constants
+
+# pylint: disable=unused-argument,missing-docstring,invalid-name
+
+class tf_op:
+    _OPSETS = collections.OrderedDict()
+    _MAPPING = None
+
+    def __init__(self, name, domain=constants.ONNX_DOMAIN, **kwargs):
+        if not isinstance(name, list):
+            name = [name]
+        self.name = name
+        self.domain = domain
+        self.kwargs = kwargs
+
+    def __call__(self, func):
+        opset = tf_op._OPSETS.get(self.domain)
+        if not opset:
+            opset = []
+            tf_op._OPSETS[self.domain] = opset
+        for k, v in inspect.getmembers(func, inspect.ismethod):
+            if k.startswith("version_"):
+                version = int(k.replace("version_", ""))
+                while version >= len(opset):
+                    opset.append({})
+                opset_dict = opset[version]
+                for name in self.name:
+                    opset_dict[name] = (v, self.kwargs)
+
+    def register_compat_handler(self, func, version):
+        opset = tf_op._OPSETS.get(self.domain)
+        if not opset:
+            opset = []
+            tf_op._OPSETS[self.domain] = opset
+            while version >= len(opset):
+                opset.append({})
+            opset_dict = opset[version]
+            opset_dict[self.name[0]] = (func, self.kwargs)
+
+    @staticmethod
+    def get_opsets():
+        return tf_op._OPSETS
+
+    @staticmethod
+    def create_mapping(max_opset, extra_opsets):
+        mapping = {constants.ONNX_DOMAIN: max_opset}
+        if extra_opsets:
+            for extra_opset in extra_opsets:
+                mapping[extra_opset.domain] = extra_opset.version
+        ops_mapping = {}
+        for domain, opsets in tf_op.get_opsets().items():
+            for target_opset, op_map in enumerate(opsets):
+                m = mapping.get(domain)
+                if m:
+                    if target_opset <= m and op_map:
+                        ops_mapping.update(op_map)
+
+        tf_op._MAPPING = ops_mapping
+        return ops_mapping
+
+    @staticmethod
+    def find_op(name):
```

Comment:
is this a must? if one knows the tf op name, he can invoke corresponding method directly.

if we decide to keep this method, maybe we should add a ```domain``` argument, to allow to filter op by domain name.

Reply:
I renamed it to find_effective_op() - it is suppose to find the op that really goes into the graph. Today we do this a little iffy, finding it with something like handler.tf_op.find_effective_op("Conv2D")  seems to be better way.